File: benchmarks/pandas/bench_period_index_methods.py
"""Benchmark: PeriodIndex.shift / sort_values / unique / to_timestamp — PeriodIndex operations on 1k periods."""
import json, time
import pandas as pd
import numpy as np

SIZE = 1_000
WARMUP = 5
ITERATIONS = 50

base = pd.Period("2020-01-01", freq="D")
shuffled = [base + ((i * 7) % SIZE) for i in range(SIZE)]
idx = pd.PeriodIndex(shuffled)

for _ in range(WARMUP):
    idx.shift(30)
    idx.sort_values()
    idx.unique()
    idx.to_timestamp(how="start")
    idx.to_timestamp(how="end")

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    idx.shift(30)
    idx.sort_values()
    idx.unique()
    idx.to_timestamp(how="start")
    idx.to_timestamp(how="end")
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function": "period_index_methods", "mean_ms": round(total_ms / ITERATIONS, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_period_index_methods.ts
/**
 * Benchmark: PeriodIndex.shift / sort / unique / toDatetimeStart / toDatetimeEnd — PeriodIndex operations on 1k periods.
 * Outputs JSON: {"function": "period_index_methods", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Period, PeriodIndex } from "../../src/index.ts";

const SIZE = 1_000;
const WARMUP = 5;
const ITERATIONS = 50;

const base = Period.fromDate(new Date(Date.UTC(2020, 0, 1)), "D");
// Build a shuffled index with some duplicates
const shuffled = Array.from({ length: SIZE }, (_, i) => base.add((i * 7) % SIZE));
const idx = PeriodIndex.fromPeriods(shuffled);

for (let i = 0; i < WARMUP; i++) {
  idx.shift(30);
  idx.sort();
  idx.unique();
  idx.toDatetimeStart();
  idx.toDatetimeEnd();
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  idx.shift(30);
  idx.sort();
  idx.unique();
  idx.toDatetimeStart();
  idx.toDatetimeEnd();
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "period_index_methods",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
